lldb/source/Plugins/TypeSystem/Swift/TypeSystemSwiftTypeRef.cpp

```diff
@@ -1422,7 +1422,26 @@ static bool ContainsSugaredParen(swift::Demangle::NodePointer node) {
 
   return false;
 }
-  
+
+swift::Demangle::NodePointer
+StripPrivateIDs(swift::Demangle::Demangler &dem,
+                swift::Demangle::NodePointer node) {
+  using namespace swift::Demangle;
+  return TypeSystemSwiftTypeRef::Transform(dem, node, [&](NodePointer node) {
+    if (node->getKind() != Node::Kind::PrivateDeclName ||
+        node->getNumChildren() != 2)
+      return node;
+
+    assert(node->getFirstChild()->getKind() == Node::Kind::Identifier);
+    assert(node->getLastChild()->getKind() == Node::Kind::Identifier);
+    auto *new_node = dem.createNode(Node::Kind::PrivateDeclName);
+    auto *ident = dem.createNodeWithAllocatedText(
+        Node::Kind::Identifier, node->getLastChild()->getText());
+    new_node->addChild(ident, dem);
+    return new_node;
```

Comment:
This function strips the private discriminators (identifiers) from a demangled symbol. This function is used when comparing symbols in `Equivalent`, so that two symbols that differ only by the unique identifier in `PrivateDeclName`, are considered equal.

Reply:
Would this hide bugs where we accidentally mix up these two types named Foo?

a.swift
`fileprivate struct Foo {}`
b.swift 
`fileprivate struct Foo { let i = "i am different" }`

Reply:
What I want to say is: we need to make sure to not forget to investigate why private discriminators reported by remote mirrors look different than the ones the Swift compiler emits in the debug info.
But that's not something blocking this patch. (But maybe there should be a bug FIXME in the code).

Reply:
good catch. I'll have to constrain this more.

Reply:
I ended up asking about this, and the explanation is:

> the private identifier is not ABI and doesn't show up in any runtime data structures. the `$` is the address of the anonymized context descriptor that represents the private scope. there are no stable string identifiers for private or local types

Reply:
With that being the case, I'm thinking that it's ok to strip private declarations for the purpose of checking equivalent types across reflection and the AST.

In other words, a symbol that has a private identifier, the runtime will never match the AST, so the only way to compare is to compare by excluding the private identifier. This does open the possibility that they're not the same type, but it seems there's no more precise way to verify.

Reply:
Up to now, I've been thinking about this only from the perspective of validation between runtime and AST. However, now I'm wondering whether we'll ever need to use a type returned by the runtime, and look it up in the AST. For most types this is fine, but for types with private identifiers, we'd have a problem.

Do you know of cases where this could happen?

Reply:
For the purpose of verification, it's okay to compare tham in a relaxed manner. But for correctness, this is a potential problem with our incremental adoption of TypeSystemSwiftTypeRef. Since we mix&match between type systems, this makes it impossible to convert between AST <-> TypeRef types with private discriminators unless we have a ValueObject with an AST type to bind the two together.

We do store the private discriminator of each source file in DWARF (as a DW_TAG_namespace). If there is a way to tie a typeref type to a DWARF compile unit, we could do the conversion from runtime -> ast discriminator. But I don't think this is going to be possible in all cases: For example when using remote mirrors to resolve the dynamic type of an existential, how would we know where that type was defined?

Reply:
Is my understanding here correct?

This is a problem for private types which lldb has the AST for. For the SDK, we'll never have private types, so the issue of AST <-> TypeRef conversion can't occur for SDK types.

For project code, there is code that's built using full swiftmodules, and code that's builds using library evolution and swiftinterfaces. For builds using full fidelity swiftmodules, can or should lldb use the AST only? If AST were strictly used when it's available, and TypeRefs used only when the AST is not available, then would we never need to worry about conversion? However, would we be able to recognize the difference between a full swiftmodule, and a swiftmodule generated from a swiftinterface?

Reply:
I'm not sure whether the TypeRef can be cross referenced to a compile unit. I'll have to look into that.

I wonder if we can cross reference through uniqueness? If, ignoring the private identifier, there's only a single type by that name (unlike say your `fileprivate struct Foo` example), then if we enumerated all types we would see that there's just the one. However that sounds heavy handed and poor for performance/maintenance.

Reply:
> Since we mix&match between type systems

do we need to do this, and if so why? To put it another way, if we get a private TypeRef, when might lldb try to use that type in an AST context, vs always using reflection for such types? In other words, once you go TypeRef, do you ever need to go back to AST?

Reply:
> Is my understanding here correct?
> 
> This is a problem for private types which lldb has the AST for. For the SDK, we'll never have private types, so the issue of AST <-> TypeRef conversion can't occur for SDK types.

No Swift ASTs are _shipped_ for the SDK overlays. Therefore, we may get private types from Remote Mirrors, but never in the ASTs. This same is not true for development builds where we build the older snapshot of the Swift overlays that is inside the apple/swift repository from source.

The situation for SDK types could arise in the testsuite but not in production.

> 
> For project code, there is code that's built using full swiftmodules, and code that's builds using library evolution and swiftinterfaces. For builds using full fidelity swiftmodules, can or should lldb use the AST only? If AST were strictly used when it's available, and TypeRefs used only when the AST is not available, then would we never need to worry about conversion? However, would we be able to recognize the difference between a full swiftmodule, and a swiftmodule generated from a swiftinterface?

Here is gets much more complicated. In the expression evaluator all type lookup requests from the Swift compiler get sent back to LLDB, and it's LLDB's job to resolve type names. LLDB might perform dynamic type resolution (using Remote Mirrors) here and then could try to reconstruct a type that does not exist in the AST world. We also can run into the exact same issue with Clang types. Dynamic type resolution can return a private implementation subclass (e.g., NSString has multiple internal implementations) that the AST cannot know about because its declaration is not part of the SDK at all.
In the case of private subclasses, we could try reconstructing more general types (superclasses) until we find one that can be reconstructed. But this does not work for all private types in general. LLDB would probably need to make up a declaration for these types and hand it to the AST.